fuse_publishers/src/graph_dot_writer.cpp

```diff
@@ -0,0 +1,203 @@
+    : uuid_map_(uuid_map), type_map_(type_map), subtype_map_(subtype_map)
+  {
+  }
+
+  template <class Vertex>
+  void operator()(std::ostream& os, const Vertex& v) const
+  {
+    os << "["
+       << "shape=" << TYPE_SHAPES[type_map_[v]] << ", label=\"" << uuid_map_[v] << "\""
+       << ", taillabel=\"" << subtype_map_[v] << "\""
+       << "]";
+  }
+
+private:
+  UUIDMap uuid_map_;
+  TypeMap type_map_;
+  SubTypeMap subtype_map_;
+
+  static const char* const TYPE_SHAPES[];
+};
+
+template <class UUIDMap, class TypeMap, class SubTypeMap>
+const char* const vertex_writer<UUIDMap, TypeMap, SubTypeMap>::TYPE_SHAPES[] = { "circle", "diamond", "none" };
+
+template <class UUIDMap, class TypeMap, class SubTypeMap>
+inline vertex_writer<UUIDMap, TypeMap, SubTypeMap> make_vertex_writer(UUIDMap uuid_map, TypeMap type_map,
+                                                                      SubTypeMap subtype_map)
+{
+  return vertex_writer<UUIDMap, TypeMap, SubTypeMap>(uuid_map, type_map, subtype_map);
+}
+
+class GraphDOTWriter
+{
+public:
+  using Edge = std::pair<std::string, std::string>;
+  using Graph = boost::adjacency_list<boost::vecS, boost::vecS, boost::undirectedS, VertexProperty>;
+
+  GraphDOTWriter() = delete;
+  ~GraphDOTWriter() = delete;
+
+  static Graph toBoost(const fuse_msgs::GraphConstPtr& msg);
+
+  static std::ostream& write(std::ostream& os, const fuse_msgs::GraphConstPtr& msg);
+};
+
+GraphDOTWriter::Graph GraphDOTWriter::toBoost(const fuse_msgs::GraphConstPtr& msg)
+{
+  using VertexMap = std::map<std::string, Graph::vertex_descriptor>;
+
+  Graph graph;
+
+  VertexMap variable_vertex_by_uuid;
+  for (const auto& variable : msg->variables)
+  {
+    variable_vertex_by_uuid[variable.uuid] = boost::add_vertex(variable, graph);
+  }
+
+  for (const auto& constraint : msg->constraints)
+  {
+    const auto constraint_vertex = boost::add_vertex(constraint, graph);
+
+    for (const auto& variable : constraint.variables)
+    {
+      boost::add_edge(constraint_vertex, variable_vertex_by_uuid[variable], graph);
+    }
+  }
+
+  return graph;
+}
+
+std::ostream& GraphDOTWriter::write(std::ostream& os, const fuse_msgs::GraphConstPtr& msg)
+{
+  const auto graph = toBoost(msg);
+  boost::write_graphviz(os, graph,
+                        make_vertex_writer(boost::get(&VertexProperty::uuid, graph),
+                                           boost::get(&VertexProperty::type, graph),
+                                           boost::get(&VertexProperty::subtype, graph)));
+  return os;
+}
+
+class GraphDOTWriterNode
+{
+public:
+  GraphDOTWriterNode();
+
+  void graphCallback(const fuse_msgs::GraphConstPtr& msg);
+
+private:
+  ros::NodeHandle nh_;
+  ros::Subscriber subscriber_;
+
+  std::string filename_prefix_{ "" };
+  int queue_size_{ 100 };
+};
+
+GraphDOTWriterNode::GraphDOTWriterNode()
+{
+  ros::NodeHandle private_node_handle("~");
+
+  private_node_handle.param("filename_prefix", filename_prefix_, filename_prefix_);
```

Comment:
I'm going to change this to something that specifies the actual format to print the filename, so the `msg->header.seq` is `0`-padded.